README.md
# About Config::Neat

Configuration files don't have to be ugly. Inspired by
[nginx configuration files](http://wiki.nginx.org/FullExample)
I decided to go a bit further and implement a format which is highly readable,
less error prone, requires little to no special markup, and yet is robust
to allow nested blocks for the projects that require more than just a plain
key-value list.

Currently Config::Neat is implemented as a suite of Perl modules
which, in addition to parsing and rendering configuration files,
implement automatic configuration file inheritance (aka includes)
and validation against schema (see below).

### Simple Syntax Example

In its simplest form, the configuration file can look like this:

    # Server configuration
    server     Some string
    port       8080
    port       8081
    use_ssl    YES

You are not forced to enclose strings in quotes, or specify delimiters
at the end of each line; you will never need to escape single or double quotes.

### Robust Syntax Example

When it comes to having different (even nested) sections,
multiline lists or strings, block comments, Config::Neat will
offer you such an opportunity:

    /*
        Global server configuration
    */
    server {
        listen                  8080
        use_ssl                 YES
        debug                   NO
        log_format              $remote_addr - $remote_user [$time]
                                $status $size $request

        supported_mime_types    text/html text/css text/xml text/plain
                                image/gif image/jpeg image/png image/x-icon
                                application/x-javascript

        /*
            My virtual hosts
        */
        virtual_hosts {

            www.domain.com {
                root            /var/www/domain
                ...
            }

            www.otherdomain.com {
                root            /var/www/otherdomain
                ...
            }
        }
    }

### Full Syntax Example

See `sample/readme.nconf` file, which gives a full overview
of the supported syntax.

## Perl Module

Perl module is located in `perl/lib` subdirectory.
It depends on [Tie::IxHash](http://search.cpan.org/~chorny/Tie-IxHash/)
module available from CPAN.

### Synopsis

    use Config::Neat;

    my $cfg = Config::Neat->new();
    my $data = $cfg->parse_file('/path/to/myconfig.nconf');

    # now $data contains a parsed hash tree which you can examine

    # consider the example config above

    my $list = $data->{'server'}->{'supported_mime_types'};
    #
    # $list now is an array reference:
    #     ['text/html', 'text/css', ..., 'application/x-javascript']

    my $log_format = $data->{'server'}->{'log_format'}->as_string;
    #
    # $log_format now is a scalar:
    #     '$remote_addr - $remote_user [$time] $status $size $request'

## Config::Neat::Inheritable

This module adds config inheritance to Config::Neat files by automatically
processing `@inherit file#subnode`, `-somekey` and `+otherkey` keys.
See [Config::Neat::Inheritable source code](perl/lib/Config/Neat/Inheritable.pm)
for further explanation.

## Config::Neat::Schema

This module adds config validation against provided schema (schema itself
can be defined using Config::Neat format). See
[Config::Neat::Schema source code](perl/lib/Config/Neat/Schema.pm)
for further explanation.

## Config::Neat::Render

This module allows you render Config::Neat-compatible structures from your data
(but read below for limitations).

### When shoud I use it?

1. When you need to convert your old configuration files to a new format
   (and then manually tweak the output).
2. When you want to dump some data for diff purposes or just for reading.
3. When readability of your output is more important than knowing original
   data types of each node in your data output.

### When shoud I NOT use it?

Do not use it for arbitrary data serialization/desrialization.
JSON and YAML will work better for this kind of task.

Why? Because Config::Neat was primarily designed to allow easier configuration
file authoring and reading, and uses relaxed syntax where strings are treated like
space-separated arrays (and vice versa), and where there's no strict definition
for boolean types, no null values, etc.

It's the developer's responsibility to treat any given parameter as a boolean,
or string, or an array. This means that once you serialize your string into
Config::Neat format and parse it back, it will be converted to an array,
and you will need to use `->as_string` method to get the value as string.

In other words, when doing this:

    my $c = Config::Neat->new();
    my $r = Config::Neat::Render->new();
    my $parsed_data = $c->parse($r->render($arbitrary_data));

$parsed_data will almost always be different from $arbitrary_data.

However, doing this immediately after:

    my $parsed_data_2 = $c->parse($r->render($parsed_data));

Should produce the same data structure again.

### Synopsis

    use Config::Neat::Render;

    my $r = Config::Neat::Render->new();

    my $data = {
        'foo' => 'Hello, World!',
        'bar' => [1, 2, 3],
        'baz' => {
            'etc' => ['foo bar', 'baz', '', 1]
        }
    };

    print $r->render($data);

The output will be:

    bar    1 2 3

    baz
    {
        etc    `foo bar` baz `` 1
    }

    foo    Hello, World!

Note that hashes in Perl do not guarantee the correct order, so blocks may have
individual parameters shuffled randomly. To sort the keys, you can provide a reference
to an ordered list of key names in the `sort` option:

    ...

    my @order = qw(foo bar baz);

    print $r->render($data, {sort => \@order});

And now the output will be:

    foo    Hello, World!
    bar    1 2 3

    baz
    {
        etc    `foo bar` baz `` 1
    }

Alternatively, setting `sort` to a true value will just sort keys alphabetically.

## Tools

### [dump-nconf](perl/bin/dump-nconf)

This script will read the configuration file
and emit the parsed data structure in either Config::Neat::Render, Data::Dumper
or JSON format. This script can be used to validate the syntax of the
configuration file and understand its internal tree representation.

## Syntax Highlighters

There are a few syntax highlighters available (see the `highlighters` folder):

1. for [Sublime Text](http://www.sublimetext.com/) desktop editor
   (also compatible with [TextMate](http://macromates.com/))
2. for [Visual Studio Code](https://code.visualstudio.com)
3. for JavaScript-based editor called [CodeMirror](http://codemirror.net/).

You can also use CodeMirror with Config::Neat highlighter to
[statically highlight](http://codemirror.net/demo/runmode.html)
configuration snippets/examples within web pages.


perl/Changes
Revision history for Config-Neat

1.4     July 24, 2019
        - Allow rendering of keys with spaces
        - Provide better error reporting on infinite @inherit loop errors

1.302   May 25, 2017
        - Fix rendering of '0'
        - Properly render empty default values as ``

1.301   March 2, 2017
        - Fix test data initialization code and update test data itself
          to reflect changes introduced in 1.3

1.3     February 28, 2017
        - Fix issue #4 (Incorrect handling of LIST items with inheritance).
          Note that with this fix the internal raw representation of the
          arrays in the config data has changed, but the API masks that
          change and stays the same; applying a schema to a config also
          ensures that the resulting data structure is correct

        - Fix issue #5 (Validator will now die on bogus boolean values;
          previously it would only warn and continue)

1.204   October 4, 2015
        - Add `to_ixhash_recursive` helper function to Config::Neat::Util

1.203   August 5, 2015
        - Add missing 'JSON' dependency (used in `dump-nconf`)

1.202   August 5, 2015
        - Add missing test files to the distribution

1.201   August 5, 2015
        - Get rid of File::Slurp::Tiny in favor of simple read_file function

1.2     August 5, 2015

        - Fixed properly check for unmatched closing bracket, add corresponding test
        - Added check and test case for missing closing bracket
        - Fixed some warnings, mute 'unitialized' warnings where appropriate
        - Fixed parsing of multi-line comments in this form:  /** / ... /**/

1.101   October 26, 2014

        - Bugfix: store 'converted to array' flags in a context-aware stack

1.1     October 26, 2014

        - Fixed and simplified parsing for repeating nodes

        - More parsing tests; use subtests for better structuring

1.0     October 25, 2014

        - Added the support for 'LIST' schema type
          to always cast single or multiple entries with the
          same name to an array (list) of nested items

        - Replace File::Slurp with File::Slurp::Tiny
          (as per https://rt.cpan.org/Public/Bug/Display.html?id=95680)

        - is_any_hash, is_any_array convenience functions
          in Config::Neat::Util

0.9     May 11, 2014

        - Implemented proper cascaded inheritance

0.8     May 10, 2014

        - Fix inheriting the same local paths from different files
          (save the file name as a part of the cache item key)

0.7     April 11, 2014

        - Automatically convert merged node to an IxHash
          if necessary

        - Allow casting any hash to an array

        - Support rendering default hash values

        - Do not count the key length for the hash keys
          that don't have values

0.6     March 12, 2014

        - Add the minimal version requirement for Test::More (0.94)

        - Must clone node before putting it into the cache;
          added test for multiple node reuse

0.5     March 4, 2014

        - Report a proper error when trying to render objects
          of unsupported types

        - Tests no longer have any hardcoded paths in the
          reference data

0.4     March 2, 2014

        - A more correct schema-based automatic casting
          from HASH to ARRAY (+ additional test)

        - Add File::Slurp to the list of build prerequisites

0.3     February 28, 2014

        - By default, hashes are now merged (no need to set '+'
          prefix for the merged key); arrays, on the contrary,
          are replaced by default, so '+' is needed to merge
          their values

        - When merging nodes, the original order of the keys
          is preserved

        - Support for hashes with mixed explicitly set and
          automatically generated indexed keys; hashes
          with sequential numbering can be cast to ARRAY
          automatically using Config::Neat::Schema

        * Initial CPAN release *

0.2     February 23, 2014

        - Support multiple key declarations at the same level
          (these will be converted to an array).

        - Support installation via Module::Build

...


0.0     September 22, 2012

        * Code published on GitHub *


perl/t/render.pl
#!/usr/bin/perl

use strict;
use utf8;

BEGIN {
  use File::Spec::Functions qw(rel2abs);
  use File::Basename;
  unshift(@INC, dirname(rel2abs($0)).'/../lib');
}

use Config::Neat;
use Config::Neat::Render;
use Data::Dumper;

my $c = Config::Neat->new();
my $r = Config::Neat::Render->new();

my $data = {
    'word' => 'Hello',
    'string_with_spaces' => 'Hello, world!',
    'simple_array' => [1, '2 3', '4 5', 6, 7],
    'wrappable_string' => 'Vestibulum ullamcorper leo quam, vel adipiscing tellus. '.
                          'Phasellus placerat dolor sit amet lorem mattis dictum. '.
                          'Vestibulum luctus malesuada risus, et porttitor lectus lobortis et.',
    'wrappable_array' => [
        'this particular array item will not wrap since it contains spaces and thus is enclosed in backticks',
        'Aa','Bb','Cc','Dd','Ee','Ff','Gg','Hh','Ii','Jj','Kk','Ll','Mm','Nn','Oo','Pp','Qq','Rr',
        'Ss','Tt','Uu','Vv','Ww','Xx','Yy','Zz',
    ],
    'subsection_1' => {
        'etc' => 'foo bar',
        'pwd' => ['foo', 'foo bar', 'ba``z', '`'],
    },
    'subsection_2' => {
        'true_parameter'  => 1,
        'false_parameter' => undef,
        'subtree' => {
            'empty_string' => '',
            'array_of_empty_strings' => ['', '', ''],
        },
    },
    'array_of_hashes' => [
        {
            'param1' => 'value1',
        },
        {
            'param2' => 'value2',
        },
        {
            'param3' => 'value3',
        },
    ],
};

my @order = qw(
    word string_with_spaces simple_array wrappable_string
    wrappable_array subsection_1 subsection_2 array_of_hashes

    etc pwd

    true_parameter false_parameter subtree

    empty_string array_of_empty_strings
);

#my $sort = undef; # unsorted
#my $sort = 1; # sorted alphabetically
my $sort = \@order; # sort in the order

my $options = {
    sort      => $sort,
    align_all => 1,
};

my $text1 = $r->render($data, $options);
my $data1 = $c->parse($text1);
my $text2 = $r->render($data1, $options);
if ($text1 ne $text2) {
    warn "WARNING: Text from two passes differs\n";

    print "# =[1]====\n";
    print $text1;
    print "# =[2]====\n";
    print $text2;
    print "# =[1:data]=====\n";
    print "/*\n";
    print Dumper($data1);
    print "*/";
} else {
    print $text1;
}


perl/Build.PL
use 5.010_001;
use strict;
use warnings FATAL => 'all';
use Module::Build;

my $builder = Module::Build->new(
    module_name              => 'Config::Neat',
    license                  => 'perl',
    dist_author              => 'Igor Afanasyev <igor.afanasyev@gmail.com>',
    dist_version_from        => 'lib/Config/Neat.pm',
    release_status           => 'stable',
    configure_requires       => {
        'Module::Build'      => 0,
    },
    build_requires           => {
        'Test::More'         => '0.94',
    },
    requires => {
        'Tie::IxHash'        => '1.22',
        'JSON'               => 0,
    },
    add_to_cleanup           => ['Config-Neat-*', 'Makefile.PL', 'MYMETA.*', 'META.*'],
    create_makefile_pl       => 'traditional',
);

$builder->create_build_script();


perl/lib/Config/Neat/Util.pm
=head1 NAME

Config::Neat::Util - Common utility functions for other Config::Neat modules

=head1 COPYRIGHT

Copyright (C) 2012-2015 Igor Afanasyev <igor.afanasyev@gmail.com>

=head1 SEE ALSO

L<https://github.com/iafan/Config-Neat>

=cut

package Config::Neat::Util;

our $VERSION = '1.4';

use strict;

use Tie::IxHash;

our @ISA = qw(Exporter);

our @EXPORT_OK = qw(
    new_ixhash
    to_ixhash
    to_ixhash_recursive
    is_number
    is_code
    is_hash
    is_ixhash
    is_any_hash
    is_array
    is_neat_array
    is_any_array
    is_scalar
    is_simple_array
    is_homogenous_simple_array
    hash_has_only_sequential_keys
    hash_has_sequential_keys
    get_next_auto_key
    offset_keys
    get_keys_in_order
    reorder_ixhash_numerically
    reorder_ixhash
    rename_ixhash_key
    read_file
);

sub new_ixhash {
    my $new = {};
    tie(%$new, 'Tie::IxHash');
    return $new;
}

sub to_ixhash {
    my $node = shift;
    die "Not a regular hash" unless is_hash($node) && !is_ixhash($node);
    my $new = new_ixhash;
    map { $new->{$_} = $node->{$_} } keys %$node;
    return $new;
}

sub to_ixhash_recursive {
    my ($node) = @_;
    my $result = is_hash($node) && !is_ixhash($node) ? to_ixhash($node) : $node;
    if (is_ixhash($result)) {
        map {
            $result->{$_} = to_ixhash_recursive($result->{$_});
        } keys %$result;
    }
    return $result;
}

sub is_number {
    my $n = shift;
    return defined $n && $n =~ m/^\d+$/;
}

sub is_code {
    my $node = shift;
    return ref($node) eq 'CODE';
}

sub is_hash {
    my $node = shift;
    return ref($node) eq 'HASH';
}

sub is_ixhash {
    my $node = shift;
    return undef unless is_hash($node);
    return ref(tied(%$node)) eq 'Tie::IxHash';
}

sub is_any_hash {
    my $node = shift;
    return is_hash($node) || is_ixhash($node);
}

sub is_array {
    my $node = shift;
    return ref($node) eq 'ARRAY';
}

sub is_neat_array {
    my $node = shift;
    return ref($node) eq 'Config::Neat::Array';
}

sub is_any_array {
    my $node = shift;
    return is_array($node) || is_neat_array($node);
}

sub is_scalar {
    my $node = shift;
    return (ref(\$node) eq 'SCALAR') or (ref($node) eq 'SCALAR');
}

sub is_simple_array {
    my $node = shift;

    return 1 if is_scalar($node);
    return undef unless is_array($node) || is_neat_array($node);

    foreach my $value (@$node) {
        return undef unless is_scalar($value);
    }
    return 1;
}

sub is_homogenous_simple_array {
    my $node = shift;

    return 1 if is_scalar($node);
    return undef unless is_array($node) || is_neat_array($node);

    my $contains_hash = undef;
    my $contains_array = undef;
    my $contains_scalar = undef;

    foreach my $value (@$node) {
        if (is_hash($value)) {
            $contains_hash |= 1;
        } elsif (is_any_array($value)) {
            $contains_array |= 1;
        } else {
            $contains_scalar |= is_scalar($value);
        }
        die "Mixing hashes with simple arrays/scalars within one node is not supported" if $contains_hash && $contains_scalar;
    }
    return $contains_scalar && !$contains_array;
}

sub hash_has_only_sequential_keys {
    my $node = shift;
    return hash_has_sequential_keys($node, 1);
}

sub hash_has_sequential_keys {
    my ($node, $strict) = @_;
    die "Not a hash" unless is_hash($node);

    my $i = 0;
    map {
        if (is_number($_)) {
            return undef if $_ != $i;
            $i++;
        } else {
            return undef if $strict;
        }
    } keys %$node;
    return 1;
}

# supposed to be used against hash that matches the
# `hash_has_sequential_keys() == true` criterion
sub get_next_auto_key {
    my $node = shift;
    die "Not a hash" unless is_hash($node);

    # get max(key)
    my $i = -1; # so that next key will start with 0
    map {
        $i = $_ if is_number($_) && $_ > $i;
    } keys %$node;

    # return max + 1
    return $i + 1;
}

sub offset_keys {
    my ($node, $offset) = @_;
    die "Not a Tie::IxHash" unless is_ixhash($node);
    return $node if $offset == 0;

    my $result = new_ixhash;

    # remap keys
    map {
        if (is_number($_)) {
            $result->{$_ + $offset} = $node->{$_};
        } else {
            $result->{$_} = $node->{$_};
        }
    } keys %$node;

    return $result;
}

# accepts an array of hasrefs
sub get_keys_in_order {
    my $result = new_ixhash;

    map {
        map {
            $result->{$_} = 1;
        } keys %$_;
    } @_;

    return keys %$result;
}

sub reorder_ixhash_numerically {
    my ($node) = @_;
    die "Not a Tie::IxHash" unless is_ixhash($node);

    # sort keys numerically
    my @a = sort {$a <=> $b} keys %$node;

    return reorder_ixhash($node, \@a);
}

sub reorder_ixhash {
    my ($node, $keysref) = @_;
    die "Not a Tie::IxHash" unless is_ixhash($node);

    my $result = new_ixhash;
    map { $result->{$_} = $node->{$_} if exists $node->{$_} } @$keysref;

    return $result;
}

sub rename_ixhash_key {
    my ($node, $from, $to) = @_;
    die "Not a Tie::IxHash" unless is_ixhash($node);
    die "Can\'t rename key '$from' to '$to', because the target key already exists" if exists $node->{$to};

    my $result = new_ixhash;
    map {
        my $key = $_ eq $from ? $to : $_;
        $result->{$key} = $node->{$_};
    } keys %$node;

    return $result;
}

sub read_file {
    my ($filename, $binmode) = @_;

    open(CFG, $filename) or die "Can't open [$filename]: $!";
    binmode(CFG, $binmode || ':utf8');
    my $text = join('', <CFG>);
    close(CFG);

    return $text;
} # end sub


1;

perl/lib/Config/Neat/Array.pm
=head1 NAME

Config::Neat::Array - Class for array-like config nodes

=head1 COPYRIGHT

Copyright (C) 2012-2015 Igor Afanasyev <igor.afanasyev@gmail.com>

=head1 SEE ALSO

L<https://github.com/iafan/Config-Neat>

=cut

package Config::Neat::Array;

our $VERSION = '1.4';

use strict;

no warnings qw(uninitialized);

use Config::Neat::Util qw(is_any_array is_neat_array);

sub new {
    my ($class, $self) = @_;
    $self = [] unless defined $self && ref($self) eq 'ARRAY';
    bless $self, $class;
    return $self;
}

sub push {
    my $self = shift;
    push @$self, @_;
}

# return a flattened one-dimensional array, where nested
# Config::Neat arrays are expanded recursively
sub as_flat_array {
    my ($self) = @_;

    # fist check if conversion will be needed
    my $need_conversion;
    foreach my $val (@$self) {
        if (is_neat_array($val)) {
            $need_conversion = 1;
            last;
        }
    }
    return $self unless $need_conversion;

    # flatten the array recursively
    my $result = Config::Neat::Array->new();
    foreach my $val (@$self) {
        if (is_neat_array($val)) {
            $val = $val->as_flat_array;
        }

        if (is_any_array($val)) {
            # expand arrays
            push @$result, @$val;
        } else {
            #push scalars and hashes as is
            push @$result, $val;
        }
    }
    return $result;
}

# Given ['foo', 'bar', 'baz'] as the contents of the array, returns 'foo bar baz' string.
# Array is flattened before being converted into a string.
# If string starts from a newline and the next line is indented, remove that amount of spaces
# from each line and trim leading and trailing newline
sub as_string {
    my ($self) = @_;

    my $val = join(' ', @{$self->as_flat_array});
    my $indent = undef;
    while ($val =~ m/\n(\s+)/g) {
        my $len = length($1);
        $indent = $len unless defined $indent and $len > 0;
        $indent = $len if $len > 0 and $indent > $len;
    }
    if ($indent > 0) {
        $indent = ' ' x $indent;
        $val =~ s/\n$indent/\n/sg;
        $val =~ s/^\s*\n//s; # remove first single newline and preceeding whitespace
        $val =~ s/\n\s*$//s; # remove last single newline and whitespace after it
    }
    return $val;
} # end sub

# Returns true if the string representation of the array
# evaluates case-insensitively to a known list of positive boolean strings
sub as_boolean {
    my ($self) = @_;

    return ($self->as_string =~ m/^(YES|Y|ON|TRUE|1)$/i);
} # end sub

# Returns true if the string representation of the array
# evaluates case-insensitively to a known list of positive or negative boolean strings
sub is_boolean {
    my ($self) = @_;

    return ($self->as_string =~ m/^(YES|NO|Y|N|ON|OFF|TRUE|FALSE|1|0)$/i);
} # end sub

# Given ['foo', 'bar', 'baz'] as the contents of the array,
# and property name 'x', returns the following hash reference:
#   {
#       0 => {'x' => 'foo'},
#       1 => {'x' => 'bar'},
#       2 => {'x' => 'baz'}
#   }
sub as_hash {
    my ($self, $propname) = @_;

    die "Second parameter (propname) not provided" unless defined $propname;

    my $result = {};
    tie(%$result, 'Tie::IxHash');

    my $n = 0;
    foreach my $val (@$self) {
        $result->{$n++} = {$propname => $val};
    }

    return $result;
} # end sub

1;


perl/lib/Config/Neat.pm
=head1 NAME

Config::Neat - Parse/render human-readable configuration files with inheritance and schema validation

=head1 SYNOPSIS

    use Config::Neat;

    my $cfg = Config::Neat->new();
    my $data = $cfg->parse_file('server.nconf');

=head1 DESCRIPTION

This module provides parsing capabilites for the Config::Neat configuration file
format (see the example below). This is a highly readable and clean format inspired
by [nginx configuration files](http://wiki.nginx.org/FullExample).
See L<https://github.com/iafan/Config-Neat/blob/master/sample/readme.nconf>
for the detailed file syntax specification.

Sample configuration file (let's call it 'server.nconf'):

    # Server configuration

    host                    localhost
    port                    8080
    use_ssl                 YES
    supported_mime_types    text/html text/css text/xml text/plain
                            image/gif image/jpeg image/png image/x-icon
                            application/x-javascript

    handler test1 {
        url                 /test1
        class               MyApp::Test
    }

    handler test2 {
        url                 /test2
        class               MyApp::AnotherTest
    }

Sample usage:

    use Config::Neat;

    my $cfg = Config::Neat->new();
    my $data = $cfg->parse_file('server.nconf');

    # now $data contains a parsed hash tree which you can examine

    # consider the sample configuration file above

    my $list = $data->{'server'}->{'supported_mime_types'};
    #
    # $list now is an array reference:
    #     ['text/html', 'text/css', ..., 'application/x-javascript']

    my $handlers = $data->{'handler'};
    map {
        print $_->{url}->as_string, ' maps to ', $_->{class}->as_string
    } @$handlers;

=head1 COPYRIGHT

Copyright (C) 2012-2015 Igor Afanasyev <igor.afanasyev@gmail.com>

=head1 SEE ALSO

L<https://github.com/iafan/Config-Neat>

=cut

package Config::Neat;

our $VERSION = '1.4';

use strict;

use Config::Neat::Array;
use Config::Neat::Util qw(is_neat_array new_ixhash get_next_auto_key read_file);
use Tie::IxHash;

my $LINE_START    = 0;
my $KEY           = 1;
my $WHITESPACE    = 2;
my $VALUE         = 3;
my $LINE_COMMENT  = 4;
my $BLOCK_COMMENT = 5;

#
# Initialize object
#
sub new {
    my ($class) = @_;

    my $self = {
        cfg => {}
    };

    bless $self, $class;
    return $self;
}

# Given a string representation of the config, returns a parsed tree
sub parse {
    my ($self, $nconf) = @_;

    my $o = {
        context            => [new_ixhash],
        context_data       => [{}],
        c                  => undef,

        pos                => 0,

        key                => '',
        values             => Config::Neat::Array->new(),
        value              => undef,
        mode               => $LINE_START,
        previous_mode      => $LINE_START,
        was_backslash      => undef,
        was_slash          => undef,
        was_asterisk       => undef,
        first_value_pos    => 0,
    };

    my $in_raw_mode     = undef;
    my $line            = 1;

    sub end_of_param {
        my ($o, $no_default_param) = @_;

        if ($o->{key} ne '') {
            push @{$o->{values}}, 'YES' if !$no_default_param && scalar(@{$o->{values}}) == 0;
            my $current_ctx = $o->{context}->[$#{$o->{context}}];
            my $data = $o->{context_data}->[$#{$o->{context_data}}];
            if (exists $current_ctx->{$o->{key}}) {
                $data->{is_array} = {} unless exists $data->{is_array};
                if (!$data->{is_array}->{$o->{key}}) {
                    $current_ctx->{$o->{key}} = Config::Neat::Array->new([$current_ctx->{$o->{key}}]);
                    $data->{is_array}->{$o->{key}} = 1;
                }
                $current_ctx->{$o->{key}}->push($o->{values});
            } else {
                $current_ctx->{$o->{key}} = $o->{values};
            }
            $o->{values} = Config::Neat::Array->new();
            $o->{key} = '';
        }
    }

    sub append_text {
        my ($o, $text) = @_;

        if ($o->{mode} == $LINE_START) {
            if (($o->{first_value_pos} > 0) and ($o->{pos} >= $o->{first_value_pos})) {
                $o->{mode} = $VALUE;
            } else {
                end_of_param($o);
                $o->{mode} = $KEY;
                $o->{first_value_pos} = 0;
            }
        } elsif ($o->{mode} == $WHITESPACE) {
            $o->{mode} = $VALUE;
            if ($o->{first_value_pos} == 0) {
                $o->{first_value_pos} = $o->{pos} - 1; # -1 to allow for non-hanging backtick before the first value
            }
        }

        if ($o->{mode} == $KEY) {
            $o->{key} .= $text;
        } elsif ($o->{mode} == $VALUE) {
            $o->{value} .= $text;
        } else {
            die "Unexpected mode $o->{mode}";
        }
    }

    sub process_pending_chars {
        my $o = shift;

        if ($o->{was_slash}) {
            append_text($o, '/');
            $o->{was_slash} = undef;
        }

        if ($o->{was_backslash}) {
            append_text($o, '\\');
            $o->{was_backslash} = undef;
        }
    }

    sub process_char {
        my $o = shift;

        process_pending_chars($o);

        append_text($o, $o->{c});
        $o->{c} = undef;
    }

    sub end_of_value {
        my $o = shift;

        process_pending_chars($o);

        if (defined $o->{value}) {
            push @{$o->{values}}, $o->{value};
            $o->{value} = undef;
        }
    }

    for (my $i = 0, my $l = length($nconf); $i < $l; $i++) {
        my $c = $o->{c} = substr($nconf, $i, 1);
        $o->{pos}++;

        if ($c ne '/') {
            $o->{was_asterisk} = undef;
        }

        if ($c eq '{') {
            next if ($o->{mode} == $LINE_COMMENT) or ($o->{mode} == $BLOCK_COMMENT);

            if ($in_raw_mode) {
                process_char($o);
                next;
            }

            end_of_value($o);

            if (!$o->{key}) {
                $o->{key} = get_next_auto_key($o->{context}->[$#{$o->{context}}]);
            }

            my $old_values = $o->{values};
            my $new_context = $o->{values} = new_ixhash;

            end_of_param($o, 1); # do not push a default parameter

            $o->{value} = undef;
            $o->{mode} = $LINE_START;
            $o->{first_value_pos} = 0;

            push @{$o->{context}}, $new_context;
            push @{$o->{context_data}}, {};

            # any values preceding the block will be added into it with an empty key value
            if (scalar(@{$old_values}) > 0) {
                $new_context->{''} = $old_values;
            }

        } elsif ($c eq '}') {
            next if ($o->{mode} == $LINE_COMMENT) or ($o->{mode} == $BLOCK_COMMENT);

            if ($in_raw_mode) {
                process_char($o);
                next;
            }

            end_of_value($o);
            end_of_param($o);

            if (scalar(@{$o->{context}}) == 1) {
                die "Unmatched closing bracket at config line $line position $o->{pos}";
            }
            pop @{$o->{context}};
            pop @{$o->{context_data}};
            $o->{mode} = $WHITESPACE;
            $o->{key} = '';
            $o->{values} = Config::Neat::Array->new();

        } elsif ($c eq '\\') {
            next if ($o->{mode} == $LINE_COMMENT) or ($o->{mode} == $BLOCK_COMMENT);

            process_pending_chars($o);

            $o->{was_backslash} = 1; # do not print current slash, but wait for the next char
            next;

        } elsif ($c eq '/') {
            next if ($o->{mode} == $LINE_COMMENT);
            next if (!$o->{was_asterisk} and $o->{mode} == $BLOCK_COMMENT);

            if ($in_raw_mode) {
                process_char($o);
                next;
            }

            if ($o->{was_asterisk} and ($o->{mode} == $BLOCK_COMMENT)) {
                $o->{mode} = $o->{previous_mode};
                next;
            }

            process_pending_chars($o);

            $o->{was_slash} = 1; # do not print current slash, but wait for the next char
            next;

        } elsif ($c eq '*') {
            next if ($o->{mode} == $LINE_COMMENT);

            if ($o->{mode} == $BLOCK_COMMENT) {
                $o->{was_asterisk} = 1;
                next;
            } else {
                if ($o->{was_slash}) {
                    $o->{was_slash} = undef;
                    $o->{previous_mode} = $o->{mode};
                    $o->{mode} = $BLOCK_COMMENT;
                    next;
                }

                process_char($o);
            }

        } elsif ($c eq '`') {
            next if ($o->{mode} == $LINE_COMMENT) or ($o->{mode} == $BLOCK_COMMENT);

            if ($o->{was_backslash}) {
                $o->{was_backslash} = undef;
                process_char($o);
                next;
            }

            $o->{c} = '';
            process_char($o);

            $in_raw_mode = !$in_raw_mode;

        } elsif (($c eq ' ') or ($c eq "\t")) {
            if ($c eq "\t") {
                warn "Tab symbol at config line $line position $o->{pos} (replace tabs with spaces to ensure proper parsing of multiline parameters)";
            }

            next if ($o->{mode} == $LINE_COMMENT) or ($o->{mode} == $BLOCK_COMMENT);

            if ($in_raw_mode) {
                process_char($o);
                next;
            }

            if ($o->{mode} == $KEY) {
                $o->{mode} = $WHITESPACE;
            } elsif ($o->{mode} == $VALUE) {
                end_of_value($o);
                $o->{mode} = $WHITESPACE;
            }

        } elsif ($c eq "\r") {
            next;

        } elsif ($c eq "\n") {
            $line++;
            $o->{pos} = 0;

            next if ($o->{mode} == $BLOCK_COMMENT);

            if ($in_raw_mode) {
                process_char($o);
                next;
            }

            end_of_value($o);
            $o->{mode} = $LINE_START;

        } elsif ($c eq "#") {
            next if ($o->{mode} == $LINE_COMMENT) or ($o->{mode} == $BLOCK_COMMENT);

            if ($in_raw_mode) {
                process_char($o);
                next;
            }

            if (($o->{mode} == $LINE_START) or ($o->{mode} == $WHITESPACE)) {
                $o->{mode} = $LINE_COMMENT;
            } else {
                process_char($o);
            }

        } else {
            next if ($o->{mode} == $LINE_COMMENT) or ($o->{mode} == $BLOCK_COMMENT);

            process_char($o);
        }

        $o->{was_asterisk} = undef;
    }

    die "Unmatched backtick at config line $line position $o->{pos}" if $in_raw_mode;

    die "Missing closing bracket at config line $line position $o->{pos}" if @{$o->{context}} > 1;

    end_of_value($o);
    end_of_param($o);

    return $self->{cfg} = $o->{context}->[0];
} # end sub

# Given file name, will read this file in the specified mode (defaults to UTF-8) and parse it
sub parse_file {
    my ($self, $filename, $binmode) = @_;
    return $self->parse(read_file($filename, $binmode));
} # end sub

1; # return true